src/test/java/com/datalogics/pdf/samples/printing/PrintPdfTest.java

```diff
@@ -111,4 +138,55 @@ private void savePageImage(final BufferedImage image, final int pageIndex) throw
             }
         }
     }
+
+    /*
+     * TestMultiPagePrinterJob implements a 'fake' PrinterJob to intercept print requests.
+     */
+    private static class TestMultiPagePrinterJob extends FakePrinterJob {
+        /*
+         * Print the document.
+         */
+        @Override
+        public void print() throws PrinterException {
+            // Create a BufferedImage to render into
+            final int width = (int) (format.getImageableWidth() - format.getImageableX());
+            final int height = (int) (format.getImageableHeight() - format.getImageableY());
+            // NOTE: We use a TYPE_4BYTE_ABGR because it is guaranteed to use a single contiguous
+            // image data buffer. This lets us checksum the raw data for the entire image.
+            final BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_4BYTE_ABGR);
+
+            Graphics2D gfx2d = image.createGraphics();
+
+            final int[] pageIndexes = { 0, 0, 1, 1 };
+
+            for (final int pageIndex : pageIndexes) {
+                if (painter.print(gfx2d, format, pageIndex) == Printable.NO_SUCH_PAGE) {
+                    break;
+                }
+                processPageImage(image, pageIndex);
+
+                // painter.print() disposed of the Graphics2D, obtain a new one
+                gfx2d = image.createGraphics();
+                gfx2d.clearRect(0, 0, width, height);
+            }
+        }
+
+        @Override
+        public void processPageImage(final BufferedImage image, final int pageIndex) throws PrinterIOException {
+            savePageImage(image, pageIndex);
+            assertThat(image, bufferedImageHasChecksum(PAGE_IMAGE_CHECKSUMS[pageIndex]));
+        }
+
+        // Based on a code snippet from the Java tutorials:
+        // https://docs.oracle.com/javase/tutorial/2d/images/saveimage.html
+        private void savePageImage(final BufferedImage image, final int pageIndex) throws PrinterIOException {
+            final String outputName = String.format(RENDERED_IMAGE_NAME, pageIndex);
```

Comment:
I see two things here:

1. We're using the same output file name as the 'default' print test, so we'll end up overwriting the output from that test. For the purposes of evaluating the image checksum, it doesn't matter because we do that before it gets overwritten, but we'd really like to preserve the outputs from each test.

1. Since we're 'double calling' the print logic—which is the whole point of the test—we're actually writing each page _twice_..._within the same test_. Note that we're calling `savePageImage()` with a `pageIndex` of 0 twice, so it will write the output PNG once, and then overwrite the same file the next time its called. I'm less worried about this—the behavior we _want_ from the real printer is each page, one time, so maybe our virtual printer should do the same—but I want us to be clear that this is happening.